File: apps/cli/commenti_test.go
package main

import (
	"encoding/json"
	"os"
	"path/filepath"
	"reflect"
	"testing"
)

type casoConformita struct {
	Nome       string   `json:"nome"`
	Body       string   `json:"body"`
	Atteso     string   `json:"atteso"`
	Segnaposti []string `json:"segnaposti"`
}

// #643: la fixture è la fonte di verità condivisa con client TS, MCP e Rust.
func casiConformita(t *testing.T) []casoConformita {
	t.Helper()
	p := filepath.Join("..", "..", "packages", "shared-schema", "fixtures", "commenti-conformita.json")
	b, err := os.ReadFile(p)
	if err != nil {
		t.Fatalf("lettura fixture: %v", err)
	}
	var f struct {
		Casi []casoConformita `json:"casi"`
	}
	if err := json.Unmarshal(b, &f); err != nil {
		t.Fatalf("parsing fixture: %v", err)
	}
	if len(f.Casi) <= 10 {
		t.Fatalf("fixture sospettosamente corta: %d casi", len(f.Casi))
	}
	return f.Casi
}

func TestRimuoviCommentiConformita(t *testing.T) {
	for _, c := range casiConformita(t) {
		t.Run(c.Nome, func(t *testing.T) {
			if got := rimuoviCommenti(c.Body); got != c.Atteso {
				t.Errorf("rimuoviCommenti(%q) = %q, atteso %q", c.Body, got, c.Atteso)
			}
			got := estraiSegnaposti(c.Body)
			if len(got) == 0 && len(c.Segnaposti) == 0 {
				return
			}
			if !reflect.DeepEqual(got, c.Segnaposti) {
				t.Errorf("estraiSegnaposti(%q) = %v, atteso %v", c.Body, got, c.Segnaposti)
			}
		})
	}
}

func TestRimuoviCommentiIdempotente(t *testing.T) {
	for _, c := range casiConformita(t) {
		if got := rimuoviCommenti(c.Atteso); got != c.Atteso {
			t.Errorf("%s: non idempotente: %q → %q", c.Nome, c.Atteso, got)
		}
	}
}

func TestCompilaTogliCommenti(t *testing.T) {
	body := "{{!-- {{x}} qui non conta --}}\nCiao {{nome}} {{!-- inline --}}"
	got := compila(body, map[string]string{"x": "NO", "nome": "Anna"})
	if got != "Ciao Anna " {
		t.Errorf("compila = %q", got)
	}
}

func TestEspandiGlobaliNonTonaNeiCommenti(t *testing.T) {
	// L'ordine nel comando render è: commenti → globali → segnaposti.
	body := rimuoviCommenti("{{!-- {{global a}} --}}\nFirma: {{global a}}")
	got := espandiGlobali(body, map[string]string{"a": "Roberto"})
	if got != "Firma: Roberto" {
		t.Errorf("espandiGlobali = %q", got)
	}
}

File: apps/client/src/lib/commenti.test.ts
import { describe, it, expect } from "vitest";
import { intervalliCommenti, rimuoviCommenti } from "./commenti";
import { compila, estraiSegnaposti } from "./template";
// #643: tabella di conformità condivisa con MCP server, Rust e Go CLI.
// Un caso nuovo si aggiunge nella fixture, non qui.
import fixture from "../../../../packages/shared-schema/fixtures/commenti-conformita.json";

interface CasoConformita {
  nome: string;
  body: string;
  atteso: string;
  segnaposti: string[];
}

const casi: CasoConformita[] = fixture.casi;

describe("rimuoviCommenti — conformità cross-linguaggio (#643)", () => {
  it("la fixture non è vuota", () => {
    expect(casi.length).toBeGreaterThan(10);
  });

  for (const caso of casi) {
    it(caso.nome, () => {
      expect(rimuoviCommenti(caso.body)).toBe(caso.atteso);
      expect(
        estraiSegnaposti(caso.body)
          .filter((s) => !s.globale)
          .map((s) => s.nome),
      ).toEqual(caso.segnaposti);
    });
  }
});

describe("rimuoviCommenti", () => {
  it("è idempotente", () => {
    for (const caso of casi) {
      expect(rimuoviCommenti(caso.atteso)).toBe(caso.atteso);
    }
  });
});

describe("intervalliCommenti", () => {
  it("nessun commento → nessun intervallo", () => {
    expect(intervalliCommenti("Ciao {{nome}}")).toEqual([]);
  });

  it("ritorna [from, to) del token intero", () => {
    const body = "a {{!-- x --}} b";
    expect(intervalliCommenti(body)).toEqual([{ from: 2, to: 14 }]);
    expect(body.slice(2, 14)).toBe("{{!-- x --}}");
  });

  it("intervalli multipli in ordine", () => {
    const r = intervalliCommenti("{{!-- a --}}\n{{!-- b --}}");
    expect(r).toEqual([
      { from: 0, to: 12 },
      { from: 13, to: 25 },
    ]);
  });

  it("commento non chiuso → nessun intervallo", () => {
    expect(intervalliCommenti("{{!-- aperto")).toEqual([]);
  });
});

describe("compila con commenti", () => {
  it("il commento non arriva nel testo compilato", () => {
    expect(
      compila("{{!-- tono B --}}\nCiao {{nome}}", { nome: "Anna" }),
    ).toBe("Ciao Anna");
  });

  it("un segnaposto citato solo nel commento non viene compilato né lasciato", () => {
    expect(compila("{{!-- {{x}} --}}\nok", { x: "NO" })).toBe("ok");
  });

  it("un segnaposto globale nel commento non viene compilato", () => {
    expect(compila("{{!-- {{global a}} --}}\nok", {}, { a: "NO" })).toBe(
      "ok",
    );
  });

  it("commento inline dopo un segnaposto", () => {
    expect(compila("Ciao {{nome}} {{!-- nota --}}", { nome: "Anna" })).toBe(
      "Ciao Anna ",
    );
  });
});
